File: client/common/courier.go
package common

import (
	"encoding/binary"
	"net"
)

const BYTES_SIZE_MESSAGE_TYPE = 4
const BYTES_SIZE_PAYLAOD_SIZE = 4

type Courier struct {
	conn *ConnectionHandler
}

func NewCourier(id string, conn net.Conn) *Courier {
	Courier := &Courier{
		conn: NewConnectionHandler(conn),
	}
	return Courier
}

func (c *Courier) SendMessage(message_type string, message string) error {
	typeToSend := make([]byte, BYTES_SIZE_MESSAGE_TYPE)
	copy(typeToSend, []byte(message_type))

	log.Debugf("action: send_message_type | result: in_process | msg: %v",
		typeToSend,
	)

	err := c.conn.send(typeToSend)

	log.Debugf("action: send_message_type | result: success | msg: %v",
		typeToSend,
	)

	if err != nil {
		log.Debugf("action: send_message_type | result: failure | msg: %v",
			typeToSend,
		)
		return err
	}

	payload := []byte(message)
	sizePayload := len(payload)
	sizePayloadBytes := make([]byte, BYTES_SIZE_PAYLAOD_SIZE)
	binary.BigEndian.PutUint32(sizePayloadBytes, uint32(sizePayload))

	finalPayload := append(sizePayloadBytes, payload...)

	log.Debugf("action: send_message | result: in_process | msg: %v",
		finalPayload,
	)

	err = c.conn.send(finalPayload)

	if err != nil {
		log.Debugf("action: send_message | result: failure | msg: %v",
			finalPayload,
		)
		return err
	}

	log.Debugf("action: send_message | result: success | msg: %v",
		finalPayload,
	)

	return nil
}

func (c *Courier) RecvTypeMessage() (string, error) {
	msgTypeBytes, err := c.conn.recv(BYTES_SIZE_MESSAGE_TYPE)
	if err != nil {
		return "", err
	}

	return string(msgTypeBytes), nil
}

func (c *Courier) RecvPayloadMessage() (string, error) {
	sizePayloadBytes, err := c.conn.recv(BYTES_SIZE_PAYLAOD_SIZE)
	if err != nil {
		return "", err
	}

	sizePayload := int(binary.BigEndian.Uint32(sizePayloadBytes))

	payloadBytes, err := c.conn.recv(sizePayload)

	if err != nil {
		return "", err
	}

	return string(payloadBytes), nil
}

File: server/common/courier.py
import logging
from common.connectionHandler import ConnectionHandler
from common.constants import BYTES_SIZE_MESSAGE_TYPE, BYTES_SIZE_PAYLAOD_SIZE

class Courier:
    def __init__(self, client_sock):
        self._addr = client_sock.getpeername()
        self._conn = ConnectionHandler(client_sock)
        
    def sendResponseMessage(self, message_type: str):
        type_to_send = bytearray(BYTES_SIZE_MESSAGE_TYPE)
        type_to_send[:len(message_type)] = message_type.encode('utf-8')
        self._conn.send(type_to_send)
        
    def sendMessageWithPayload(self, message_type: str, message_payload: str):
        self.sendResponseMessage(message_type)
        payload = message_payload.encode('utf-8')
        size_payload = len(payload)
        size_payload_bytes = size_payload.to_bytes(BYTES_SIZE_PAYLAOD_SIZE, byteorder='big', signed=False)
        final_payload = size_payload_bytes + payload
        
        self._conn.send(final_payload)
        
        
    def recvTypeMessage(self) -> str:
        msg_type_bytes = self._conn.recv(BYTES_SIZE_MESSAGE_TYPE)        
        return msg_type_bytes.rstrip(b'\x00').decode('utf-8')
        
    def recvPayloadMessage(self) -> str:
        size_payload_bytes = self._conn.recv(BYTES_SIZE_PAYLAOD_SIZE)
        size_payload = int.from_bytes(size_payload_bytes, byteorder='big', signed=False)
        
        payload_bytes = self._conn.recv(size_payload)
        return payload_bytes.decode('utf-8')
    
    def close(self):
        self._conn.close() 